### ISSUE TITLE
prod_bw passes padded_shape bounds to ttnn::slice (which resolves logically) and compares logical_shape against padded_shape in its guard

### ISSUE THREAD

**Issue by botify-service** [tenstorrent/tt-metal#53284]
### Component / Area

ttnn eltwise unary backward — `ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp`, `prod_bw` (the `dim`-specified path, `all_dimensions == false`)

### Issue Type (optional)

Bad Outputs

### Observed

`prod_bw` mixes logical and padded shapes in two places on the same path. I want to be precise
about which half I can prove from source and which half I cannot, so they are separated below.

**1. The guard compares a logical shape against a padded shape.**

```cpp
// unary_backward.cpp:1682
if (prod_result.logical_shape() != grad.padded_shape()) {
```

These are different quantities by definition, so the comparison does not test what it reads as
testing. For a tile-aligned `grad` they happen to coincide and the test behaves; for any `grad`
whose logical extent is not a tile multiple they differ purely because of padding. Given
`prod_result` comes from `ttnn::prod(input, dim, keepdim=true)`, its logical shape carries a `1`
in the reduced dimension, so in the reducing case the two are essentially never equal and the
branch is taken unconditionally — the guard is not selecting anything. Either way it is comparing
the wrong pair; the honest reading is that one side should be `grad.logical_shape()`.

**2. Padded coordinates are handed to `ttnn::slice`, which resolves against the logical shape.**

```cpp
// dim == 3 / -1
ttsl::SmallVector<uint32_t> end_index = {
    grad.padded_shape()[0], 1, grad.padded_shape()[1], grad.padded_shape()[2]};
Tensor new_slice_tensor = ttnn::slice(required, start_index, end_index, step, std::nullopt);

// dim == 2 / -2 — same shape of bug
ttsl::SmallVector<uint32_t> end_index = {
    grad.padded_shape()[0], 1, grad.padded_shape()[1], grad.padded_shape()[3]};
```

`ttnn::slice` takes its bounds against `input_tensor.logical_shape()` (`slice.cpp:95`) and maps
them through `wrap_index(index, input_shape[i])` with `input_shape` logical
(`slice.cpp:222-223`). `wrap_index` is

```cpp
// data_movement/common/common.cpp:486
uint32_t wrap_index(int index, int size) { return index < 0 ? size + index : index; }
```

which normalises negative indices only — a positive bound past the logical extent passes through
unchanged. The sole downstream assertion is `modified_ends[i] >= modified_begins[i]`
(`slice.cpp:344`); there is no upper-bound check. The output extent is computed as `end - begin`,
so a padded bound silently yields an oversized slice that reads tile padding as data rather than
raising.

This is the same mechanism as #53283 (`split_tensor_for_glu`), in a different file and a
different op — a fix to either does not reach the other.

### What I could not determine

I have not established the exact numeric output. `prod_bw` chains `permute` → `slice` →
`permute` → `bcast`, and the trailing `bcast(reciprocal_input, temp, MUL, W/H)` may or may not
re-broadcast over the oversized extent in a way that masks or compounds it. Predicting the final
tensor would mean modelling those four ops together, which I am not willing to assert without
running it.

So the claim here is narrower than my other reports: **the bounds handed to `slice` are padded
where `slice` expects logical, and the guard compares mismatched shape kinds.** Both are visible
in the source and neither depends on hardware. Whether that surfaces as wrong gradient values, a
wrong shape, or is absorbed downstream is exactly what a run would settle, and I would rather say
so than guess.

### Expected

`prod_bw` should use `grad.logical_shape()` for the slice bounds and compare like against like in
the guard, matching `torch.prod`'s backward semantics, which are defined on logical extents with
no notion of tile padding.

```cpp
-    if (prod_result.logical_shape() != grad.padded_shape()) {
+    if (prod_result.logical_shape() != grad.logical_shape()) {
...
-        end_index = {grad.padded_shape()[0], 1, grad.padded_shape()[1], grad.padded_shape()[2]};
+        end_index = {grad.logical_shape()[0], 1, grad.logical_shape()[1], grad.logical_shape()[2]};
```

and the same substitution in the `dim == 2 / -2` branch.

### Steps to reproduce

I do not have a Tenstorrent card. These are the commands for someone who does. The shapes matter:
the last two dimensions must not be tile multiples, so that logical and padded differ.

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

for shape in [(1, 1, 50, 100), (1, 1, 32, 100), (1, 1, 32, 128)]:   # last one is the control
    x = torch.rand(shape, dtype=torch.bfloat16) + 0.5      # keep away from zero, prod_bw divides
    g = torch.randn(shape, dtype=torch.bfloat16)
    xt = x.clone().float().requires_grad_(True)
    torch.prod(xt, dim=3, keepdim=True).backward(
        torch.prod(xt, dim=3, keepdim=True).detach() * 0 + g.float().sum(dim=3, keepdim=True))
    tx = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    tg = ttnn.from_torch(g, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    out = ttnn.to_torch(ttnn.from_device(ttnn.prod_bw(tg, tx, dim=3)[0]))
    print(f"{shape}: ttnn {tuple(out.shape)} vs input {tuple(x.shape)} "
          f"{'ok' if out.shape == x.shape else 'SHAPE MISMATCH'}")

ttnn.close_device(device)
```

The gradient of `prod` w.r.t. its input has the input's shape, so a shape mismatch on the first two
rows and a match on `(1, 1, 32, 128)` would confirm the padded-bounds path. If the shapes all
match, the slice is being absorbed downstream and the report reduces to the guard being a
type confusion — still worth fixing, but not a bad-output bug.

### Why the current tests do not catch it

- The `prod_bw` sweeps and unit tests use tile-aligned shapes, so `logical_shape() ==
  padded_shape()` throughout and both defects are inert.
- #14886 ("ttnn.prod_bw - Low PCC", closed) exercised accuracy on aligned shapes; nothing samples
  a width or height between tile boundaries.

### Is this a regression?

No, and I did not bisect. Both reads are present at the SHA below. Context rather than blame: the
same logical/padded confusion appears in `_variance_impl` (#53277) and `split_tensor_for_glu`
(#53283), which suggests a shared habit from before the logical/padded distinction was
consistently drawn, rather than three unrelated slips.

### Software Versions

| | |
|---|---|
| tt-metal | `main`, source as of 2026-08-16 |
| reference | `torch.prod` backward semantics (logical extents) |

### Hardware Details

No Tenstorrent card and no ttsim run.

The two source facts — padded bounds into a logical-resolving `slice`, and the mismatched-kind
comparison — are read directly from `unary_backward.cpp`, `slice.cpp` and `common.cpp` at the
lines quoted. The non-clamping behaviour of `wrap_index` and the absence of an upper-bound assert
in `slice` I verified while investigating #53283. The downstream consequence is explicitly left
open above rather than asserted.

### Impact

`prod_bw` is not on a hot training path for most models, so I would not rate this urgent. It is
filed because the mechanism is confirmed and shared with #53283, and because a gradient that is
silently wrong on some shapes and right on others is the kind of defect that costs a great deal of
time to attribute if it does bite.
